controllers/propagator/template_utils.go

```diff
@@ -0,0 +1,641 @@
+}
+
+type templateResolverWithCancel struct {
+	// cancel will stop the TemplateResolver and the automatic token refresh.
+	cancel context.CancelFunc
+	// canceled will be set if resolver failed to start or was stopped. The atomic.Bool type is used to avoid using
+	// a write lock.
+	canceled atomic.Bool
+	// referenceCount is a counter for how many replicated policies rely on this template resolver. Once it is 0,
+	// the template resolver will be cleaned up. A pointer is used since the reference count can be copied to a new
+	// templateResolverWithCancel instance if the TemplateResolver instance was unexpectedly canceled and was
+	// restarted.
+	referenceCount *atomic.Uint32
+	// The template resolver instance. If the global lock was acquired, this will be set as long as canceled is false.
+	resolver *templates.TemplateResolver
+}
+
+// TemplateResolvers handles managing the template resolvers that policies leverage, including the default template
+// resolver. This also handles garbage collection of unused template resolvers except for the default template resolver.
+// This is based on a reference count, so it's important to call RemovePolicy when a replicated policy is deleted.
+// GetResolver handles all other reference count updates. All methods are concurrency safe, however, write lock
+// contention could occur if there a large amount of replicated policies with spec.hubTemplateOptions.serviceAccountName
+// reference a nonexistent service account since errors encountered instantiating the template resolver are not cached
+// so that it can be retried.
+type TemplateResolvers struct {
+	// The main context used for all child contexts. Canceling this context will cause TemplateResolvers to shutdown.
+	contollerCtx context.Context //nolint:containedctx
+	// A client used to generate tokens through the TokenRequst API.
+	client *kubernetes.Clientset
+	// The Kubernetes configuration used as the base for template resolver clients. No authentication information is
+	// copied, only configuration identifying the server and TLS configuration is used.
+	config *rest.Config
+	// The controller-runtime client to leverage the cache of policy objects.
+	mgrClient client.Client
+	// A wait group to wait for clean up once controllerCtx is canceled.
+	wg *sync.WaitGroup
+	// A channel to trigger replicated policy reconciles if an asynchronous error occurs such as refreshing the token
+	// failing.
+	replicatedPolicyUpdates chan event.GenericEvent
+	// globalLock has a write lock if a template resolver is being instantiated or getting cleaned up. All other
+	// operations use a read lock.
+	globalLock sync.RWMutex
+	// The default template resolver when a replicated policy does not specify a service account.
+	defaultTemplateResolver *templates.TemplateResolver
+	// A map of service acconts to template resolvers. Access to this must leverage globalLock.
+	TemplateResolvers map[types.NamespacedName]*templateResolverWithCancel
```

Comment:
types.NamespacedName namespace is always rootPlc's namespace right?

Reply:
For this map, the key is the service account the template resolver uses.

Reply:
Oh now I got it rootPlc namespace could vary

Reply:
Yes, exactly.

Reply:
One more question, should the ServiceAccount be created in the same namespace as the root policy??

Reply:
@yiraeChristineKim that's right. Only service accounts of the root policy namespace can be used.